Not work on CentOS 6.8

version: libkqueue-2.3.1 release

To build using cmake 3.6.1 on CentOS 6.8, temporarily modify the `cmake_minimum_required(VERSION 3.8.0)` to `cmake_minimum_required(VERSION 3.6.1)`. (It won't have any bad effect since we don't need to build rpm package here.)

```bash
$ mkdir build
$ cd build/
$ cmake -DENABLE_TESTING=YES ..
-- The C compiler identification is GNU 7.3.1
-- Check for working C compiler: /opt/rh/devtoolset-7/root/usr/bin/cc
-- Check for working C compiler: /opt/rh/devtoolset-7/root/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Looking for include file sys/signalfd.h
-- Looking for include file sys/signalfd.h - found
-- Looking for include file sys/timerfd.h
-- Looking for include file sys/timerfd.h - found
-- Looking for include file sys/eventfd.h
-- Looking for include file sys/eventfd.h - found
-- Looking for include file err.h
-- Looking for include file err.h - found
-- Looking for EPOLLRDHUP
-- Looking for EPOLLRDHUP - found
-- Looking for NOTE_TRUNCATE
-- Looking for NOTE_TRUNCATE - not found
-- Looking for NOTE_REVOKE
-- Looking for NOTE_REVOKE - not found
-- Looking for ppoll
-- Looking for ppoll - found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vagrant/libkqueue-2.3.1/build
$ make
Scanning dependencies of target kqueue
[  4%] Building C object CMakeFiles/kqueue.dir/src/common/filter.c.o
[  8%] Building C object CMakeFiles/kqueue.dir/src/common/kevent.c.o
[ 12%] Building C object CMakeFiles/kqueue.dir/src/common/knote.c.o
[ 16%] Building C object CMakeFiles/kqueue.dir/src/common/kqueue.c.o
[ 20%] Building C object CMakeFiles/kqueue.dir/src/common/map.c.o
[ 25%] Building C object CMakeFiles/kqueue.dir/src/posix/platform.c.o
[ 29%] Building C object CMakeFiles/kqueue.dir/src/linux/platform.c.o
[ 33%] Building C object CMakeFiles/kqueue.dir/src/linux/read.c.o
[ 37%] Building C object CMakeFiles/kqueue.dir/src/linux/signal.c.o
[ 41%] Building C object CMakeFiles/kqueue.dir/src/linux/timer.c.o
[ 45%] Building C object CMakeFiles/kqueue.dir/src/linux/user.c.o
[ 50%] Building C object CMakeFiles/kqueue.dir/src/linux/vnode.c.o
[ 54%] Building C object CMakeFiles/kqueue.dir/src/linux/write.c.o
[ 58%] Linking C shared library libkqueue.so
[ 58%] Built target kqueue
Scanning dependencies of target libkqueue-test
[ 62%] Building C object test/CMakeFiles/libkqueue-test.dir/kevent.c.o
[ 66%] Building C object test/CMakeFiles/libkqueue-test.dir/main.c.o
[ 70%] Building C object test/CMakeFiles/libkqueue-test.dir/read.c.o
[ 75%] Building C object test/CMakeFiles/libkqueue-test.dir/test.c.o
[ 79%] Building C object test/CMakeFiles/libkqueue-test.dir/timer.c.o
[ 83%] Building C object test/CMakeFiles/libkqueue-test.dir/user.c.o
[ 87%] Building C object test/CMakeFiles/libkqueue-test.dir/vnode.c.o
[ 91%] Building C object test/CMakeFiles/libkqueue-test.dir/proc.c.o
[ 95%] Building C object test/CMakeFiles/libkqueue-test.dir/signal.c.o
[100%] Linking C executable libkqueue-test
[100%] Built target libkqueue-test
$ make test
Running tests...
Test project /home/vagrant/libkqueue-2.3.1/build
    Start 1: libkqueue-test
1/1 Test #1: libkqueue-test ...................***Failed    0.00 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.01 sec

The following tests FAILED:
	  1 - libkqueue-test (Failed)
Errors while running CTest
make: *** [Makefile:85：test] 错误 8
$ $ ./test/libkqueue-test 
Running 1 iterations
1: test_peer_close_detection()	
2: test_kqueue()	
3: test_kevent()	
4: test_cleanup()	
5: test_ev_receipt()	
6: test_kevent_socket_add()	
Unable to add the following kevent:
[ident=4, filter=-1, flags = 1 (EV_ADD), fflags = 0, data=0, udata=(nil)]
kevent_add(): kevent: Bad address
***** ERROR: Program received signal 6 *****
[0x0]
[0x0]
[0x7ffceca97530]
/lib64/libc.so.6(_IO_2_1_stderr_+0x0)[0x388898f120]
[0x7ffceca97530]
[0xfbad2086]
[0x7ffceca97540]
/lib64/libc.so.6(_IO_vfprintf+0x24e)[0x38886440fe]
./test/libkqueue-test[0x406067]
[0x1]
[0x0]
/lib64/libc.so.6(_IO_2_1_stdout_+0x0)[0x388898f040]
[0x7ffceca97580]
[0x1c]
[0x7ffceca97590]
./test/libkqueue-test[0x4068f5]
[0x7ffceca97420]
/lib64/libc.so.6(_IO_vfprintf+0x1fe)[0x38886440ae]
./test/libkqueue-test[0x406040]
[0x1]
[0x1]
[0x0]
[0x7ffceca9748c]
[0x7ffceca974a8]
[0x0]
[0x7ffceca974c8]
[0x7ffceca97400]
[0x15]
./test/libkqueue-test[0x4068ef]
./test/libkqueue-test[0x4068f5]
[0xffffffff00000000]
[0x0]
 *** TEST FAILED ***
```

--- a/src/linux/platform.c
+++ b/src/linux/platform.c
@@ -677,24 +677,27 @@ linux_get_descriptor_type(struct knote *kn)
         kn->kn_flags |= KNFL_SOCKET_PASSIVE;
 
 #ifdef SO_GET_FILTER
-    /*
-     * Test if socket has a filter
-     * pcap file descriptors need to be considered as passive sockets as
-     * SIOCINQ always returns 0 even if data is available.
-     * Looking at SO_GET_FILTER is a good way of doing this.
-     */
-    socklen_t out_len = 0;
-    ret = getsockopt(fd, SOL_SOCKET, SO_GET_FILTER, NULL, &out_len);
-    if (ret < 0) {
-        switch (errno) {
-            case ENOTSOCK:   /* same as lsock = 0 */
-                break;
-            default:
-                dbg_perror("getsockopt(3)");
-                return (-1);
-        }
-    } else if (out_len)
-        kn->kn_flags |= KNFL_SOCKET_PASSIVE;
+    {
+        socklen_t out_len = 0;
+        
+        /*
+         * Test if socket has a filter
+         * pcap file descriptors need to be considered as passive sockets as
+         * SIOCINQ always returns 0 even if data is available.
+         * Looking at SO_GET_FILTER is a good way of doing this.
+         */
+        ret = getsockopt(fd, SOL_SOCKET, SO_GET_FILTER, NULL, &out_len);
+        if (ret < 0) {
+            switch (errno) {
+                case ENOTSOCK:   /* same as lsock = 0 */
+                    break;
+                default:
+                    dbg_perror("getsockopt(3)");
+                    return (-1);
+            }
+        } else if (out_len)
+            kn->kn_flags |= KNFL_SOCKET_PASSIVE;
+    }
 #endif /* SO_GET_FILTER */
 
     return (0);

--- a/src/linux/platform.c
+++ b/src/linux/platform.c
@@ -20,9 +20,6 @@
 #include <sys/resource.h>
 #include "../common/private.h"
 
-#ifndef SO_GET_FILTER
-#define SO_GET_FILTER SO_ATTACH_FILTER
-#endif
 
 //XXX-FIXME TEMP
 const struct filter evfilt_proc = EVFILT_NOTIMPL;
@@ -581,7 +578,6 @@ linux_get_descriptor_type(struct knote *kn)
     socklen_t slen;
     struct stat sb;
     int ret, lsock, stype;
-    socklen_t out_len;
     const int fd = (int)kn->kev.ident;
 
     /*
@@ -680,13 +676,14 @@ linux_get_descriptor_type(struct knote *kn)
     } else if (lsock)
         kn->kn_flags |= KNFL_SOCKET_PASSIVE;
 
+#ifdef SO_GET_FILTER
     /*
      * Test if socket has a filter
      * pcap file descriptors need to be considered as passive sockets as
      * SIOCINQ always returns 0 even if data is available.
      * Looking at SO_GET_FILTER is a good way of doing this.
      */
-    out_len = 0;
+    socklen_t out_len = 0;
     ret = getsockopt(fd, SOL_SOCKET, SO_GET_FILTER, NULL, &out_len);
     if (ret < 0) {
         switch (errno) {
@@ -698,6 +695,7 @@ linux_get_descriptor_type(struct knote *kn)
         }
     } else if (out_len)
         kn->kn_flags |= KNFL_SOCKET_PASSIVE;
+#endif /* SO_GET_FILTER */
 
     return (0);
 }

